mypyc/genops.py

```diff
@@ -775,36 +775,51 @@ def gen_continue(self, builder: 'IRBuilder', line: int) -> None: pass
     def gen_return(self, builder: 'IRBuilder', value: Value, line: int) -> None: pass
 
 
-class BaseNonlocalControl(NonlocalControl):
+class LoopNonlocalControl(NonlocalControl):
+    def __init__(self, outer: NonlocalControl,
+                 continue_block: BasicBlock, break_block: BasicBlock) -> None:
+        self.outer = outer
+        self.continue_block = continue_block
+        self.break_block = break_block
+
     def gen_break(self, builder: 'IRBuilder', line: int) -> None:
-        assert False, "break outside of loop"
+        builder.add(Goto(self.break_block))
 
     def gen_continue(self, builder: 'IRBuilder', line: int) -> None:
-        assert False, "continue outside of loop"
+        builder.add(Goto(self.continue_block))
 
     def gen_return(self, builder: 'IRBuilder', value: Value, line: int) -> None:
-        builder.add(Return(value))
+        self.outer.gen_return(builder, value, line)
 
 
-class CleanupNonlocalControl(NonlocalControl):
-    """Abstract nonlocal control that runs some cleanup code. """
-    def __init__(self, outer: NonlocalControl) -> None:
-        self.outer = outer
+class TryFinallyNonlocalControl(NonlocalControl):
```

Comment:
I'd put `TryFinallyNonlocalControl` down below Cleanup, I think. And leave `BaseNonlocalControl` first,